lib/ansible/modules/unarchive.py

```diff
@@ -1144,29 +1173,12 @@ def main():
     # Run only if we found differences (idempotence) or diff was missing
     if res_args.get('diff', True) and not module.check_mode:
         # do we need to change perms?
-        top_folders = []
-        for filename in handler.files_in_archive:
+        for filename in handler.files_created_by_archive:
             file_args['path'] = os.path.join(b_dest, to_bytes(filename, errors='surrogate_or_strict'))
-
-            try:
-                res_args['changed'] = module.set_fs_attributes_if_different(file_args, res_args['changed'], expand=False)
-            except OSError as ex:
-                module.fail_json("Unexpected error when accessing exploded file.", exception=ex, **res_args)
-
-            if '/' in filename:
-                top_folder_path = filename.split('/')[0]
-                if top_folder_path not in top_folders:
-                    top_folders.append(top_folder_path)
-
-        # make sure top folders have the right permissions
-        # https://github.com/ansible/ansible/issues/35426
-        if top_folders:
-            for f in top_folders:
-                file_args['path'] = "%s/%s" % (dest, f)
-                try:
-                    res_args['changed'] = module.set_fs_attributes_if_different(file_args, res_args['changed'], expand=False)
-                except OSError as ex:
-                    module.fail_json("Unexpected error when accessing exploded file.", exception=ex, **res_args)
+            diff_attrs = {}
+            res_args['changed'] = module.set_fs_attributes_if_different(file_args, res_args.get('changed', False), diff_attrs, expand=False)
+            if diff_attrs:
+                res_args.setdefault('diff', {})[filename] = diff_attrs
```

Comment:
@pkingstonxyz I was mulling over this suggestion a little more, and I realized the diff is programmatically available, but not displayed when `--diff` is passed on the CLI or `diff: True` is set on a task. The callback only displays known keys (like 'before'/'after' or 'prepared')  https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/callback/__init__.py#L432.

I'm also not sure whether this is complimentary or inconsistent with the diff format returned by ZipArchive.is_unarchived, and I think more opinions are needed. If you want to remove this bit, I'd merge it as-is. Sorry for the misleading suggestion.

```suggestion
            res_args['changed'] = module.set_fs_attributes_if_different(file_args, res_args.get('changed', False), expand=False)
```

Reply:
Ah okay. I went ahead and removed the diff_attrs stuff, but I'll make a note and take a deeper look at the diff format and diff behavior of this module.